### PR TITLE
[Task] Adding Bundle Config Event in Installer

### LINKED ISSUE #132: [Improvement] activate PimcoreSimpleBackendSearchBundle by default
alternative approach see https://github.com/pimcore/pimcore/issues/14110

### PR DESCRIPTION
**Pull request by mattamon**
## Additional info
Resolves https://github.com/pimcore/skeleton/issues/132

### <samp>🤖 Generated by Copilot at b2dca0d</samp>

This pull request adds a new event `BundleSetupEvent` to the installer bundle, which allows developers to customize the bundle installation process. The event can be used to add or remove bundles or recommendations for the installer, or to clear them for demos or custom setups. The pull request also updates the installer command, user interface, and documentation to support the new event.

### <samp>🤖 Generated by Copilot at b2dca0d</samp>

> _`BundleSetupEvent`_
> _Customize installer flow_
> _Autumn leaves falling_

### <samp>🤖 Generated by Copilot at b2dca0d</samp>

*  Add a new event `BundleSetupEvent` to customize the bundle installation process ([link](https://github.com/pimcore/pimcore/pull/15160/files?diff=unified&w=0#diff-426790dbb366c950e0c883528623194c392da1d66f0a7e963f234d1b42ebefecR1-R65), [link](https://github.com/pimcore/pimcore/pull/15160/files?diff=unified&w=0#diff-04236d6d406bbfbde0a2614a1971e7ec452273d9f0111ef34df29c982a521b59R1-R16), [link](https://github.com/pimcore/pimcore/pull/15160/files?diff=unified&w=0#diff-4eee5188dd424c74c2215a7595b21d4bde4443e3732f5b86d71484393076c498L29-R36), [link](https://github.com/pimcore/pimcore/pull/15160/files?diff=unified&w=0#diff-4eee5188dd424c74c2215a7595b21d4bde4443e3732f5b86d71484393076c498L60-R63), [link](https://github.com/pimcore/pimcore/pull/15160/files?diff=unified&w=0#diff-4eee5188dd424c74c2215a7595b21d4bde4443e3732f5b86d71484393076c498R193-R197), [link](https://github.com/pimcore/pimcore/pull/15160/files?diff=unified&w=0#diff-b2d6459dddd923201b6eba5251a26962ca98cb62a12750986a9223f5bb908c06L20-R22), [link](https://github.com/pimcore/pimcore/pull/15160/files?diff=unified&w=0#diff-b2d6459dddd923201b6eba5251a26962ca98cb62a12750986a9223f5bb908c06L307-R314), [link](https://github.com/pimcore/pimcore/pull/15160/files?diff=unified&w=0#diff-af9a0ec0242a3ba90ff43200e6381f96a4a69c2795a676ab53dedd6f268460abL65-R120), [link](https://github.com/pimcore/pimcore/pull/15160/files?diff=unified&w=0#diff-375aded12dfb868b84d5b8c27d1aed2358680a27b683e8a985f1b85193dd79aaR26-R28))
*  Add a new installable bundle `PimcoreTinymceBundle` that provides a rich text editor for Pimcore ([link](https://github.com/pimcore/pimcore/pull/15160/files?diff=unified&w=0#diff-4eee5188dd424c74c2215a7595b21d4bde4443e3732f5b86d71484393076c498L29-R36), [link](https://github.com/pimcore/pimcore/pull/15160/files?diff=unified&w=0#diff-4eee5188dd424c74c2215a7595b21d4bde4443e3732f5b86d71484393076c498L69-R75))
*  Update the `InstallCommand` class to dispatch the `BundleSetupEvent` and use the event-driven bundle list and recommendations for the choice question ([link](https://github.com/pimcore/pimcore/pull/15160/files?diff=unified&w=0#diff-b2d6459dddd923201b6eba5251a26962ca98cb62a12750986a9223f5bb908c06L307-R314), [link](https://github.com/pimcore/pimcore/pull/15160/files?diff=unified&w=0#diff-b2d6459dddd923201b6eba5251a26962ca98cb62a12750986a9223f5bb908c06R427-R436))
*  Replace the constant `Installer::EVENT_NAME_STEP` with the constant `InstallEvents::EVENT_NAME_STEP` to use the new class that defines the installer event names ([link](https://github.com/pimcore/pimcore/pull/15160/files?diff=unified&w=0#diff-b2d6459dddd923201b6eba5251a26962ca98cb62a12750986a9223f5bb908c06L365-R371), [link](https://github.com/pimcore/pimcore/pull/15160/files?diff=unified&w=0#diff-4eee5188dd424c74c2215a7595b21d4bde4443e3732f5b86d71484393076c498L60-R63), [link](https://github.com/pimcore/pimcore/pull/15160/files?diff=unified&w=0#diff-4eee5188dd424c74c2215a7595b21d4bde4443e3732f5b86d71484393076c498L244-R253))

### CODE REVIEW

File: bundles/InstallBundle/src/Command/InstallCommand.php [hunk 1]
```diff
@@ -418,4 +419,13 @@ private function generateBundleDescription(): string
     {
         return implode(',', array_keys(Installer::INSTALLABLE_BUNDLES));
     }
+
+    private function getRecommendBundles() {
```
Comment (blankse) [inline bundles/InstallBundle/src/Command/InstallCommand.php]:
```suggestion
    private function getRecommendBundles(): string
    {
```

File: bundles/InstallBundle/src/Event/BundleSetupEvent.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        return $this->recommendations;
+    }
+
+    public function addBundle(string $key, string $class) : void
+    {
+        $this->bundles[$key] = $class;
+    }
+
+    public function addRecommendation(string $recommendedBundleKey) {
```
Comment (blankse) [inline bundles/InstallBundle/src/Event/BundleSetupEvent.php]:
```suggestion
    public function addRecommendation(string $recommendedBundleKey): void
    {
```

**Comment by mattamon** [comment]
@mcassier31 could you please check the docs :)

File: doc/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md [hunk 3]
```diff
@@ -23,6 +23,9 @@
 #### [Installer] :
 
 -  Changed the return type of `Pimcore\Extension\Bundle\Installer\InstallerInterface::getOutput` to `BufferedOutput | NullOutput`.
+-  Adding `BundleSetupEvent` Event. The installation bundles can be customized in the installing process via an Eventlistener or EventSubscriber.
```
Comment (mcassier31) [inline doc/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md]: Not sure about "The installation bundles"?

File: doc/01_Getting_Started/02_Advanced_Installation_Topics.md [hunk 4]
```diff
@@ -62,6 +62,61 @@ Available bundles for installation:
 - PimcoreWordExportBundle (for import/export functionality for translations in Word format)
 - PimcoreXliffBundle (for import/export functionality for translations in Xliff format)
 
+#### Adding or Removing Bundles / Bundle Recommendations
+Before bundles are displayed in the installation process, the `BundleSetupEvent` is fired.
+You can listen or subscribe to this event to add/remove bundles or the recommendations.
+Note that a recommendation will only be added if the bundle is already in the bundles list.
+For more info check e.g. the [pimcore/skeleton](https://github.com/pimcore/skeleton) on how the [pimcore/admin-ui-classic-bundle](https://github.com/pimcore/admin-ui-classic-bundle) is installed there.
```
Comment (mcassier31) [inline doc/01_Getting_Started/02_Advanced_Installation_Topics.md]:
```suggestion
For more info, you can take a look at the [Pimcore Skeleton](https://github.com/pimcore/skeleton) to see how the [Admin UI Classic Bundle](https://github.com/pimcore/admin-ui-classic-bundle) is installed.
```

File: doc/01_Getting_Started/02_Advanced_Installation_Topics.md [hunk 5]
```diff
@@ -62,6 +62,61 @@ Available bundles for installation:
...
+
+    public function bundleSetup(BundleSetupEvent $event): void
+    {
+        $event->addBundle('PimcoreAdminBundle', PimcoreAdminBundle::class);
+        $event->addRecommendation('PimcoreAdminBundle');
+    }
+}
+```
+
+Make sure to register your listener/subscriber under `config/installer.yaml` like described in [Preconfiguring the Installer](#preconfiguring-the-installer).
```
Comment (mcassier31) [inline doc/01_Getting_Started/02_Advanced_Installation_Topics.md]:
```suggestion
Make sure to register your listener/subscriber under `config/installer.yaml` as described in [Preconfiguring the Installer](#preconfiguring-the-installer).
```

File: doc/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md (hunk 3 above)
Comment (mattamon) [reply doc/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md]: Change it to "Bundles that are installable"

Comment (mcassier31) [reply doc/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md]: Or "that are available for installation"?

File: bundles/InstallBundle/src/Event/BundleSetupEvent.php [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    public function addBundle(string $key, string $class) : void
+    {
+        $this->bundles[$key] = $class;
+    }
+
+    public function removeBundle(string $key): void
+    {
+        if(array_key_exists($key, $this->bundles)) {
+            unset($this->bundles[$key]);
+        }
```
Comment (jdreesen) [inline bundles/InstallBundle/src/Event/BundleSetupEvent.php]: I don't think this check is necessary. You can unset non-existent keys without an error/notice.

```suggestion
        unset($this->bundles[$key]);
```

File: bundles/InstallBundle/src/Installer.php [hunk 7]
```diff
@@ -177,15 +187,22 @@ public function needsDbCredentials(): bool
...
 
-    public function setBundlesToInstall(array $bundlesToInstall): void
+    public function setBundlesToInstall(array $bundlesToInstall, array $availableBundles): void
     {
         // map and filter the bundles
-        $bundlesToInstall = array_filter(array_map(static function (string $bundle) {
-            return self::INSTALLABLE_BUNDLES[$bundle] ?? null;
+        $bundlesToInstall = array_filter(array_map(static function (string $bundle) use ($availableBundles) {
+            return $availableBundles[$bundle] ?? null;
         }, $bundlesToInstall));
```
Comment (jdreesen) [inline bundles/InstallBundle/src/Installer.php]:
```suggestion
        $bundlesToInstall = array_filter(array_map(
            static fn (string $bundle) => $availableBundles[$bundle] ?? null,
            $bundlesToInstall,
        ));
```

File: bundles/InstallBundle/src/Event/BundleSetupEvent.php [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    public function getRecommendedBundles(): array
+    {
+        return $this->recommendations;
+    }
+
+    public function addInstallableBundle(string $key, string $class, bool $recommend = false) : void
+    {
+        $this->bundles[$key] = $class;
+        if($recommend) {
+            $this->recommendations[] = $recommendedBundleKey;
```
Comment (dvesh3) [inline bundles/InstallBundle/src/Event/BundleSetupEvent.php]:
```suggestion
            $this->recommendations[] = $key;
```

File: bundles/InstallBundle/src/Event/BundleSetupEvent.php [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    public function addRequiredBundle(string $key, string $class): void
+    {
+        $this->bundles[$key] = $class;
+        $this->required[] = $key;
+    }
+
+    public function getInstallableBundles(array $bundles): array
+    {
+        // merge the required bundles and make sure they are unique
+        return array_unique(array_merge($this->required, $bundles));
```
Comment (dvesh3) [inline bundles/InstallBundle/src/Event/BundleSetupEvent.php]: it shows required bundle in the installable list which is unnecessary 
image

Comment (mattamon) [reply bundles/InstallBundle/src/Event/BundleSetupEvent.php]: Changed it. Now the bundle is not shown anymore.

File: bundles/InstallBundle/src/Event/BundleSetupEvent.php [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        if($recommend) {
+            $this->recommendations[] = $key;
+        }
+    }
+
+    public function removeBundle(string $key): void
+    {
+        unset($this->bundles[$key]);
+        unset($this->recommendations[$key]);
+    }
```
Comment (jdreesen) [inline bundles/InstallBundle/src/Event/BundleSetupEvent.php]: Recommendations are added without a key but removed using one.

Comment (mattamon) [reply bundles/InstallBundle/src/Event/BundleSetupEvent.php]: ah sorry my bad, I was changing so much, that I totally forgot that. I will change it. Thank you for the hint!

**Review by dvesh3 — APPROVED** [review]

**Pull request merged**